bitcoin/script/miniscript.h

```diff
@@ -726,167 +748,165 @@ struct Node {
-            }
-            case NodeType::ANDOR: {
-                auto x = subs[0]->ProduceInput(ctx, nonmal), y = subs[1]->ProduceInput(ctx, nonmal), z = subs[2]->ProduceInput(ctx, nonmal);
-                return InputResult(Choose((y.nsat + x.sat).NonCanon(), z.nsat + x.nsat, nonmal), Choose(y.sat + x.sat, z.sat + x.nsat, nonmal));
-            }
-            case NodeType::WRAP_A:
-            case NodeType::WRAP_S:
-            case NodeType::WRAP_C:
-            case NodeType::WRAP_N:
-                return subs[0]->ProduceInput(ctx, nonmal);
-            case NodeType::WRAP_D: {
-                auto x = subs[0]->ProduceInput(ctx, nonmal);
-                return InputResult(ZERO, x.sat + ONE);
-            }
-            case NodeType::WRAP_J: {
-                auto x = subs[0]->ProduceInput(ctx, nonmal);
-                // If a dissatisfaction with a nonzero top stack element exists, an alternative dissatisfaction exists.
-                // As the dissatisfaction logic currently doesn't keep track of this nonzeroness property, and thus even
-                // if a dissatisfaction with a top zero element is found, we don't know whether another one with a
-                // nonzero top stack element exists. Make the conservative assumption that whenever the subexpression is weakly
-                // dissatisfiable, this alternative dissatisfaction exists and leads to malleability.
-                return InputResult(InputStack(ZERO).Malleable(x.nsat.available != Availability::NO && !x.nsat.has_sig), x.sat);
+                case NodeType::THRESH: {
+                    std::vector<InputStack> sats = Vector(EMPTY);
+                    for (size_t i = 0; i < subres.size(); ++i) {
+                        auto& res = subres[subres.size() - i - 1];
+                        std::vector<InputStack> next_sats;
+                        next_sats.push_back(sats[0] + res.nsat);
+                        for (size_t j = 1; j < sats.size(); ++j) next_sats.push_back(Choose(sats[j] + res.nsat, std::move(sats[j - 1]) + res.sat, nonmal));
+                        next_sats.push_back(std::move(sats[sats.size() - 1]) + std::move(res.sat));
+                        sats = std::move(next_sats);
+                    }
+                    InputStack nsat = INVALID;
+                    for (size_t i = 0; i < sats.size(); ++i) {
+                        if (i != node.k) nsat = Choose(std::move(nsat), std::move(sats[i]), nonmal);
+                    }
+                    assert(node.k <= sats.size());
+                    return InputResult(std::move(nsat), std::move(sats[node.k]));
+                }
+                case NodeType::OLDER: {
+                    return InputResult(INVALID, ctx.CheckOlder(node.k) ? EMPTY : INVALID);
+                }
+                case NodeType::AFTER: {
+                    return InputResult(INVALID, ctx.CheckAfter(node.k) ? EMPTY : INVALID);
```

Comment:
If the final stack does not have a signature, then the attacker can alter the timelock values. Should that affect this line? For example, (I think) the current algorithm would mark or_i(after(9), pk(sig)) as non-malleable if timelock is not met.

Reply:
That's true, but I'd rather fix this in a different PR, as there are multiple possible solutions, and I'd prefer to keep this PR non-behavior-changing.